### ISSUE TITLE
Process backlog more slowly to avoid sucking up resources and hurting the target

### ISSUE THREAD

**Issue by otisg** [sematext/logagent-js#38]
Related to #37 

Instead of processing a potentially big backlog as fast as possible (and in the process negatively affecting other resources - CPU, network consumption, the output target, etc.) it may be better to start backlog processing slowly and then slowly increase the processing rate until the optimal speed has been achieved.

How "optical speed" is defined/determined is TBD. :)
Maybe Logagent should inspect responses it gets from the target looking for errors, and if certain errors/error codes are returned reduce the processing speed.  Maybe it should keep track of response times and slow down if they are starting to take too long.

Maybe there should be a fixed maxOutputRate, too, that could be used as the absolute max limit.

**Comment by megastef** [comment]
Please note there was a bug in logsene-js, which did not close the socket in some cases (http errors) and it did not respect the MAX_CLIENT_SOCKETS limit. I think node.js opened then too many connections and Java backends create then too many threads and start at some point to reject the HTTP requests. Anyhow, this should be fixed in logsene-js@1.1.25. 

The current algorithm for Elasticsearch/Logsene target in logsne-js:

1. As soon max number of events or max size of bulk request is reached ship logs to the target.
   Default settings: 1000 events or 3 MB bulk.  The "lastSend" date gets stored.

2. Check every 20 seconds (configurable in [logsene-js](https://github.com/sematext/logsene-js#environment-variables) ) for events to ship, and send the current buffer as bulk (if the the lastSend date was more than 20 seconds ago). This is to minimize the delay for low log volumes. 

3.  Failed bulk requests from 1) and 2) are stored to a disk queue (default max size is 10000 files). Once a minute this queue is checked for files and the files are shipped sequentially.  This might be a problem, when the queue is large - triggering e.g. 10k sequential http requests (as long all HTTP requests are successful). Failed re-transmits go back to the disk buffer. When one error happens the re-transmit operation gets stopped and will be triggered again a minute later. 

For 1) there is currently no way to reduce the http requests - if we detect an "error" state, we could send the data straight to the disk-buffer (switch to store & forward mode) until the disk-buffer is indexed successfully. 

I agree that it makes sense to watch how Elasticsearch responses to the bulk request, and slow down bulk indexing when Elasticsearch gets slower or starts to reject the data. But as more logagent slows down the consumption of disk space will grow and might hit the storage limit.

**Comment by otisg** [comment]
> But as more logagent slows down the consumption of disk space will grow and might hit the storage limit.

Right. But I think that should/could be addressed separately by adding the ability to drop the oldest data in the on-disk buffer if the aggregate buffer size exceeds N bytes.  In other words, just because Logagent currently doesn't do this doesn't mean that it should slam the destination.

**Comment by megastef** [comment]
Logagent drops the oldest data. The disk buffer is a "ring buffer" - by default with max. 10000 files (configurable). 
I mean with  "hit the storage limit"  that 10k files might be reached and old files are then lost.
